### ISSUE TITLE
Homebridge-Nest failed today after working for quite some time.

### ISSUE THREAD

**Issue by davidtennant8082** [KraigM/homebridge-nest#107]
I've had home bridge up and running for well over a year now with limited issues.  Haven't done much in the way of updating. I tried to reinstall assuming there could be an update since I last did it and the error remains. Checked with the nest developer site and checked the authorization token and set that back up incase that was causing an issue and have checked permissions there. Everything seems fine...

Here is what it is failing with currently...
XserveServer:~ Tennant$ homebridge
[1/6/2017, 11:07:38 PM] Loaded plugin: homebridge-nest
[1/6/2017, 11:07:38 PM] Registering platform 'homebridge-nest.Nest'
[1/6/2017, 11:07:38 PM] ---
[1/6/2017, 11:07:38 PM] Loaded config.json with 0 accessories and 1 platforms.
[1/6/2017, 11:07:38 PM] ---
[1/6/2017, 11:07:38 PM] Loading 1 platforms...
[1/6/2017, 11:07:38 PM] Initializing Nest platform...
[1/6/2017, 11:07:38 PM] Fetching Nest devices.
/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/node_modules/websocket-extensions/lib/websocket_extensions.js:133
    this._pipeline.processIncomingMessage(message, callback, context);
                  ^

TypeError: Cannot read property 'processIncomingMessage' of undefined
    at instance.processIncomingMessage (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/node_modules/websocket-extensions/lib/websocket_extensions.js:133:19)
    at instance._emitMessage (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:442:22)
    at instance._emitFrame (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:405:19)
    at instance.parse (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:144:18)
    at instance.parse (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/client.js:56:58)
    at IO.write (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/streams.js:80:16)
    at TLSSocket.ondata (_stream_readable.js:528:20)
    at emitOne (events.js:77:13)
    at TLSSocket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at TLSSocket.Readable.push (_stream_readable.js:110:10)
    at TLSWrap.onread (net.js:523:20)

**Comment by tobeychris** [comment]
I am seeing the same thing:
```
root@HomeBridge3:~ # homebridge
*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>
[1/7/2017, 1:59:49 AM] Loaded plugin: homebridge-nest
[1/7/2017, 1:59:49 AM] Registering platform 'homebridge-nest.Nest'
[1/7/2017, 1:59:49 AM] ---
[1/7/2017, 1:59:49 AM] Loaded config.json with 0 accessories and 1 platforms.
[1/7/2017, 1:59:49 AM] ---
[1/7/2017, 1:59:49 AM] Loading 1 platforms...
[1/7/2017, 1:59:49 AM] Initializing Nest platform...
[1/7/2017, 1:59:49 AM] Fetching Nest devices.
/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/node_modules/websocket-extensions/lib/websocket_extensions.js:133
    this._pipeline.processIncomingMessage(message, callback, context);
                  ^

TypeError: Cannot read property 'processIncomingMessage' of undefined
    at Extensions.processIncomingMessage (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/node_modules/websocket-extensions/lib/websocket_extensions.js:133:19)
    at Client._emitMessage (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:442:22)
    at Client._emitFrame (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:405:19)
    at Client.parse (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/hybi.js:144:18)
    at Client.parse (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/driver/client.js:56:58)
    at IO.write (/usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/websocket-driver/lib/websocket/streams.js:80:16)
    at TLSSocket.ondata (_stream_readable.js:555:20)
    at emitOne (events.js:96:13)
    at TLSSocket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at TLSSocket.Readable.push (_stream_readable.js:134:10)
    at TLSWrap.onread (net.js:551:20)
```

**Comment by tobeychris** [comment]
I have tried a fresh install of everything, redoing the token process, even making a new product in the Nest API.
If I remove the token then homebridge will fall-back to the legacy API (assuming credentials are there), but then you lose cameras and possibly other functionality,

**Comment by emilkwaaitaal** [comment]
Same here. Not working since today. Same error message as the others...

**Comment by davidtennant8082** [comment]
Looks like it occurred for some people under issue 106 which was closed as if just started working on its own. 
https://github.com/KraigM/homebridge-nest/issues/106

The thought was an issue on Nests end.

**Comment by tobeychris** [comment]
I agree. Definitely something on Nest's end. I guess the question is if this is temporary like #106 or a permanent change in their APIv2.

**Comment by davidtennant8082** [comment]
I had also tried a complete reinstall before posting. Have it back up and running using the legacy API for the time being.

**Comment by digitalkyle** [comment]
only the legacy api seems to be working for me as well. not sure if nest servers or whats up.

**Comment by Muzach** [comment]
Same issue here.  If Nest's v2 API has some intermittent failures it would be nice to have this js defensive so it doesn't crash homebridge...

**Comment by PencilDH** [comment]
Hello I will just add that I am having the same issue, does anyone understand why this is happening? As we all have Dev accounts we could start logging this issue with nest if this is defiantly their problem?

**Comment by pedroserano** [comment]
+1

**Comment by rtoohil** [comment]
Interestingly, I can make curl calls and get responses back just fine. I haven't dug into the code enough to understand the difference between the standard read API calls and the calls that the plugin is making, but the API itself seems fine (using the same token that I generated initially during the plugin setup).

One thing I noticed -- the error coming from faye-websocket-node is similar to this issue: https://github.com/faye/faye-websocket-node/issues/38

The firebase that is used by homebridge-nest is v2.3.2, which includes a version of faye-websocket from before that fix. I made that (one-line) [fix](https://github.com/faye/websocket-driver-node/commit/23d0a9ea5e4906155649f5106a689a884a875d7d) by hand (rather than upgrading both firebase and faye-websocket),  and homebridge started up fine and is working.

So, my guess is that a change has happened on the Nest end with how they handle closing out websocket clients. The older version of faye-websocket doesn't handle this well, returns that error, and that error isn't caught anywhere.

If I get some time later today, I'll put together a quick pull request that updates firebase to a more recent version (which will pull in a more recent version of faye-websocket), but that will require a bit of testing to make sure that the firebase methods haven't changed much between versions.

In the meantime, if you make that one line change, it should get everything back up and running (at least it has for me, at least for the last hour)

**Comment by digitalkyle** [comment]
Thank Ryan, it got ride of the error but now it won't go past the fetching nest devices. I believe the websocket is the root cause of this error

**Comment by conradhagemans** [comment]
For now I removed the Nest platform in config.json of Homebridge. Homebridge runs again fine.
Hopefully there will be an update of the NPM Nest module.

**Comment by PencilDH** [comment]
@conradhagemans You can just make it fall back to legacy API by removing the token and it will work as usual no need to remove the whole device!

**Comment by smolyneux** [comment]
I found upgrading the version of websocket-driver used (similar to @rtoohil  ’s suggestion) from **0.5.2** to **0.5.3** solved the issue for me. Running the below commands:

**cd /usr/local/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/

sudo rm -rf websocket-driver

cd ../

sudo npm install websocket-driver@0.5.3 --save**


And restarting Homebridge should allow you to continue using the newer API via a token.

**Comment by emilkwaaitaal** [comment]
Works indeed again with upgrade of websocket-driver. Thank you Smolyneux!

**Comment by davidtennant8082** [comment]
Upgrading as per smolyneux worked as well. 

I had to edit the config file as to get it to run on the legacy API, I had to remove all but the user name and password, but had just save multiple copies so that was just cut an paste the file in place and load it

**Comment by simonrb2000** [comment]
Sorry to open this up again but is this working now with the new API after updating the web socket as instructed above? I'm using the legacy API at the moment.

**Comment by davidtennant8082** [comment]
It's working with the new API for me as of right now.

**Comment by simonrb2000** [comment]
Excellent - thank you. I might update. Out of curiosity what's the difference between the old and new api within HomeKit? Using the Legacy one now for a couple of days and not seen any difference. I know the old API isn't being developed anymore but neither is the new one by looks of it. 

Cheers

**Comment by Muzach** [comment]
Thanks @smolyneux!  This worked for me too.

**Comment by simonrb2000** [comment]
I have done above steps and it worked for a couple of minutes and its not working.. Please advise if its still working your end..

Cheers

**Comment by davidtennant8082** [comment]
Still working perfectly post updating websocket-driver. 

In the troubleshooting steps originally of nuking and reinstalling everything I now have my two nest protects working with HomeKit too as I was running such an old version of the nest plugin that it only worked with the thermostat prior to the failure.

**Comment by rtoohil** [comment]
https://github.com/KraigM/homebridge-nest/pull/109

Pull request with working shrinkwrap.

**Comment by jaytay79** [comment]
I tried the websocket-driver fix but now have the same problem as @digitalkyle - homebridge won't start up as it seems to get stuck at the fetching stage. FWIW my websocket-driver module is under /usr/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/ on Ubuntu if that makes any difference....

**Comment by simonrb2000** [comment]
That's exactly the problem I had yesterday that I posted about earlier. I removed it from my config and ran homebridge overnight without it then this morning I put the next info back in the config file exactly the same and it worked?!? Bizarre - although Homebridge has crashed today though which doesn't usually happen.

On 14 Jan 2017, at 21:27, Jim <notifications@github.com<mailto:notifications@github.com>> wrote:


I tried the websocket-driver fix but now have the same problem as @digitalkyle<https://github.com/digitalkyle> - homebridge won't start up as it seems to get stuck at the fetching stage. FWIW my websocket-driver module is under /usr/lib/node_modules/homebridge-nest/node_modules/firebase/node_modules/faye-websocket/node_modules/ on Ubuntu if that makes any difference....

-
